base/traits.jl

```diff
@@ -57,3 +57,20 @@ struct RangeStepRegular   <: TypeRangeStep end # range with regular step
 struct RangeStepIrregular <: TypeRangeStep end # range with rounding error
 
 TypeRangeStep(instance) = TypeRangeStep(typeof(instance))
+
+## iterable trait
```

Comment:
This code isn't used in the PR currently, but it illustrates the alternative approach based on a trait rather than o `method_exists(start, Tuple{T})`. It fixes the type inference issue (provided the fallback doesn't call `method_exists` as it currently does).

BTW, I've noted an inconsistency in the naming of traits: we have `iteratorsize`, `iteratoreltype`, but `IndexStyle`, `TypeRangeStep`, `TypeArithmetic` and `TypeOrder`. Looks like the CamelCase variants are more numerous and more recent, so maybe we should adopt that convention everywhere? Added to #20402.

Reply:
Uppercase makes the most sense when what will be returned is a type-instance---for example, `IndexStyle(a)` will return `T()` where `T<:IndexStyle`. With better constant-prop it's less obvious that we need to return a dedicated type-instance, although that does have some advantage in clarity.

Reply:
:-1: Adding a new thing every iterable type needs to define is not ideal.

Reply:
Maybe not "ideal", but not the end of the world either IMHO given that you need to define several methods anyway. And if we really don't want to add another trait, we can add a new type to `iteratorsize` (and rename it), which will have the advantage of making the choice of the type more explicit.

Anyway I'm all ears if somebody has a better solution that works (i.e. is inferrable, see @timholy's [comment](https://github.com/JuliaLang/julia/pull/25356#discussion_r159223722) above).